zebra-consensus/src/checkpoint.rs

```diff
@@ -3,315 +3,1098 @@
+    ///
+    /// The first checkpoint does not have any ancestors, so it only verifies the
+    /// genesis block.
+    queued: BTreeMap<BlockHeight, QueuedBlockList>,
+
+    /// The range of heights that we are currently verifying. Extends from the
+    /// most recently verified checkpoint (`Excluded`), to the next highest
+    /// checkpoint (`Included`).
+    ///
+    /// If checkpoint verification has not started yet, the current range only
+    /// contains the genesis checkpoint.
+    ///
+    /// `None` means that checkpoint verification has finished.
+    current_checkpoint_range: Option<VerifyBounds>,
 }
 
-/// The error type for the CheckpointVerifier Service.
-// TODO(jlusby): Error = Report ?
-type Error = Box<dyn error::Error + Send + Sync + 'static>;
+/// The CheckpointVerifier implementation.
+///
+/// Contains non-service utility functions for CheckpointVerifiers.
+impl CheckpointVerifier {
+    /// Return a checkpoint verification service, using the provided `checkpoint_list`.
+    ///
+    /// The returned type is opaque to allow instrumentation or other wrappers, but
+    /// can be boxed for storage. It is also `Clone` to allow sharing of a
+    /// verification service.
+    ///
+    /// This function should be called only once for a particular checkpoint list (and
+    /// network), rather than constructing multiple verification services based on the
+    /// same checkpoint list.
+    //
+    // Currently only used in tests.
+    //
+    // We'll use this function in the overall verifier, which will split blocks
+    // between BlockVerifier and CheckpointVerifier.
+    #[cfg(test)]
+    fn new(
+        checkpoint_list: impl Into<BTreeMap<BlockHeight, BlockHeaderHash>>,
+    ) -> Result<Self, Error> {
+        let checkpoints: BTreeMap<BlockHeight, BlockHeaderHash> = checkpoint_list.into();
+
+        // An empty checkpoint list can't actually verify any blocks.
+        match checkpoints.keys().cloned().next() {
+            None => {
+                return Err("there must be at least one checkpoint, for the genesis block".into())
+            }
+            Some(BlockHeight(0)) => {}
+            _ => return Err("checkpoints must start at the genesis block height 0".into()),
+        };
+
+        Ok(CheckpointVerifier {
+            checkpoint_list: checkpoints,
+            queued: <BTreeMap<BlockHeight, QueuedBlockList>>::new(),
+            // We start by verifying the genesis block, by itself
+            current_checkpoint_range: Some((Included(BlockHeight(0)), Included(BlockHeight(0)))),
+        })
+    }
+
+    /// Return the block height of the highest checkpoint in the checkpoint list.
+    ///
+    /// If there is only a single checkpoint, then the maximum height will be
+    /// zero. (The genesis block.)
+    ///
+    /// The maximum height is constant for each checkpoint list.
+    ///
+    /// Returns None if there are no checkpoints.
+    fn get_max_checkpoint_height(&self) -> Option<BlockHeight> {
```

Comment:
TIOLI: I'd remove the `get_` on most of these functions, feels redundant to me

Reply:
ooh dang, 1.0 style guide.

Reply:
Fixed in 71c4b3cf90bbf8f13a398b8826bdd38002839528